File: labs_25_03_26/ex2/src/PascalTriangleRow.cpp
#include "../inc/PascalTriangleRow.h"

void PascalTriangleRow::generateRow(int n) {
  row.resize(n + 1);
  row[0] = 1;
  for (int k = 1; k <= n; k++) {
    row[k] =
        static_cast<int>(static_cast<long long>(row[k - 1]) * (n - k + 1) / k);
  }
}

PascalTriangleRow::PascalTriangleRow(int n) {
  if (n < 0) {
    throw InvalidRowNumberException("Nieprawidłowy numer wiersza");
  }
  generateRow(n);
}

int PascalTriangleRow::getElement(int m) {
  if (m < 0 || static_cast<size_t>(m) >= row.size()) {
    throw OutOfBoundsException("liczba spoza zakresu");
  }
  return row[m];
}

File: labs_26_03_30/ex1/PascalTriangleRow.java
package labs_25_03_26.ex1;

class PascalTriangleRow {
    private int[] row;

    public PascalTriangleRow(int n) throws InvalidRowNumberException {
        if (n < 0) {
            throw new InvalidRowNumberException("Nieprawidłowy numer wiersza");
        }
        generateRow(n);
    }

    private void generateRow(int n) {
        row = new int[n + 1];
        row[0] = 1;
        for (int k = 1; k <= n; k++) {
            row[k] = (int) ((long) row[k - 1] * (n - k + 1) / k);
        }
    }

    // Metoda zwracająca wartość m-tego elementu
    public int getElement(int m) throws OutOfBoundsException {
        if (m < 0 || m >= row.length) {
            throw new OutOfBoundsException("liczba spoza zakresu");
        }
        return row[m];
    }
}